### ISSUE TITLE
Append Lockdown Rule

### ISSUE THREAD

**Issue by anthonyd-cd** [cloudflare/python-cloudflare#112]
Greetings, 

I'm very sorry to bother everyone but I am very new to Python, very to programming, very new to the internet itself :|
Anyway, I am trying to simply get a list of CF Lockdown Rules, append to this list and in another Python Script, remove the last entry. Sounds easy enough, append() and pop() should do it but I can't even append yet :(

This is my code (please forgive my badly written code)

`"add_firewall_lockdown_rule.py" 21L, 692C                                                           17,4          All
#!/usr/bin/env python3
import CloudFlare
import json

def main():
    cf = CloudFlare.CloudFlare(email='admin.devops@domain.ca', token='***********************************')
    zone_id = '*******************************'
    fl = cf.zones.firewall.lockdowns.get(zone_id)
    fl_id = fl[1]['id']
    fl = fl[1]['urls']
    print(f"This is the Firewall Lockdown ID: {fl_id}")
    print(f"These are the following Lockdown Rules: \n{fl}")
    print()
    fl.append('our.domain.ca')
    new_fl = json.dumps(fl)
    print(f"These are the following Lockdown Rules: \n{new_fl}")
    cf.zones.firewall.lockdowns.put(zone_id,fl_id, new_fl)
    `
    
 I keep getting errors:

```This is the Firewall Lockdown ID: ****************************************
These are the following Lockdown Rules:
['her.domain.ca/*', 'my.domain.ca/*', 'your.domain.ca/*', 'his.domain.ca/*']

These are the following Lockdown Rules:
["her.domain.ca/*", "my.domain.ca/*", "your.domain.ca/*", "his.domain.ca/*", "our.domain.ca"]
```

Traceback (most recent call last):
  File "add_firewall_lockdown_rule.py", line 21, in <module>
    main()
  File "add_firewall_lockdown_rule.py", line 17, in main
    cf.zones.firewall.lockdowns.put(zone_id,fl_id, new_fl)
  File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 695, in put
    params, data)
  File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 128, in call_with_auth
    params, data, files)
  File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 502, in _call
    raise CloudFlareAPIError(code, message)
CloudFlare.exceptions.CloudFlareAPIError: not_found

And I am not sure how to fix it, is it expecting the ip key and value as well?

Thanks

**Comment by mahtin** [comment]
I believe the following:
```
    fl.append('our.domain.ca')
    new_fl = json.dumps(fl)
    print(f"These are the following Lockdown Rules: \n{new_fl}")
    cf.zones.firewall.lockdowns.put(zone_id, fl_id, new_fl)
```
should be:
```
    fl.append('our.domain.ca')
    new_fl = json.dumps(fl)
    print(f"These are the following Lockdown Rules: \n{new_fl}")
    cf.zones.firewall.lockdowns.put(zone_id, fl_id, fl)
```
This is because the `json.dumps()` call returns a string; however, you should just pass a native Python value (a `dict` in this case) to the put call that follows.

Please try that.

**Comment by anthonyd-cd** [comment]
Thanks for getting bak to me @mahtin, however I needed to turn into a string because when I pass the fl variable I get this error:

  File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 196, in _call_network
    url += '/' + identifier3
TypeError: must be str, not list

**Comment by anthonyd-cd** [comment]
And if I turn it into a dictionary like so: 

['her.domain.ca/*', 'my.domain.ca/*', 'your.domain.ca/*', 'his.domain.ca/*', 'our.domain.ca']
{'urls': ['her.domain.ca/*', 'my.domain.ca/*', 'your.domain.ca/*', 'his.domain.ca/*', 'our.domain.ca']}

I then get an error as well:

  File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 196, in _call_network
    url += '/' + identifier3
TypeError: must be str, not dict

**Comment by mahtin** [comment]
Ah ha! - sorry, I missed one more important part.  My bad. Here's my revised edit of your code:
```
    fl.append('our.domain.ca')
    new_fl = json.dumps(fl)
    print(f"These are the following Lockdown Rules: \n{new_fl}")
    cf.zones.firewall.lockdowns.put(zone_id, fl_id, data=fl)
```
The `data=` is very important otherwise the argument is considered to be on the API URL (vs `data`). Please look at the examples of `POST` (or `PUT) within the [README](https://github.com/cloudflare/python-cloudflare/blob/master/README.md) page.

**Comment by anthonyd-cd** [comment]
Still no luck, 

`#!/usr/bin/env python3
import CloudFlare

def main():
    cf = CloudFlare.CloudFlare(email='admin.devops@canadadrives.ca', token='71cb3f2ec47891a1dea75ad664a6f62381708')
    zone_id = 'bde285ff625c45338281577ac22d0c36'
    fl = cf.zones.firewall.lockdowns.get(zone_id)
    fl_id = fl[1]['id']
    print(f"This is the Firewall Lockdown Rule to modify: \n{fl[1]['id']}--{fl[1]['description']}")
    print()
    fl = fl[1]['urls']
    print(f"These are the Current Lockdown Rules: \n{fl}")
    print()
    fl.append('our.domain.ca')
    print(f"These are the following New Lockdown Rules: \n{fl}")
    print()
    cf.zones.firewall.lockdowns.put(zone_id, fl_id, data=fl)

    exit(0)
if __name__ == '__main__':
    main()`

Traceback (most recent call last):
  File "add_firewall_lockdown_rule_new.py", line 21, in <module>
    main()
  File "add_firewall_lockdown_rule_new.py", line 17, in main
    cf.zones.firewall.lockdowns.put(zone_id, fl_id, data=fl)
  File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 695, in put
    params, data)
  File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 128, in call_with_auth
    params, data, files)
  File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 502, in _call
    raise CloudFlareAPIError(code, message)
CloudFlare.exceptions.CloudFlareAPIError: zonelockdown.api.unknown_error

**Comment by anthonyd-cd** [comment]
Still no luck,

`#!/usr/bin/env python3
import CloudFlare

def main():
cf = CloudFlare.CloudFlare(email='admin.devops@canadadrives.ca', token='****************')
zone_id = '**************************************'
fl = cf.zones.firewall.lockdowns.get(zone_id)
fl_id = fl[1]['id']
print(f"This is the Firewall Lockdown Rule to modify: \n{fl[1]['id']}--{fl[1]['description']}")
print()
fl = fl[1]['urls']
print(f"These are the Current Lockdown Rules: \n{fl}")
print()
fl.append('our.domain.ca')
print(f"These are the following New Lockdown Rules: \n{fl}")
print()
cf.zones.firewall.lockdowns.put(zone_id, fl_id, data=fl)

if name == 'main':
main()`

Response:

This is the Firewall Lockdown Rule to modify:
**********************************--Anthony Test

These are the Current Lockdown Rules:
['her.domain.ca/*', 'my.domain.ca/*', 'your.domain.ca/*', 'his.domain.ca/*']

These are the following New Lockdown Rules:
['her.domain.ca/*', 'my.domain.ca/*', 'your.domain.ca/*', 'his.domain.ca/*', 'our.domain.ca']

Traceback (most recent call last):
File "add_firewall_lockdown_rule_new.py", line 21, in
main()
File "add_firewall_lockdown_rule_new.py", line 17, in main
cf.zones.firewall.lockdowns.put(zone_id, fl_id, data=fl)
File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 695, in put
params, data)
File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 128, in call_with_auth
params, data, files)
File "/usr/local/lib/python3.6/dist-packages/cloudflare-2.8.15-py3.6.egg/CloudFlare/cloudflare.py", line 502, in _call
raise CloudFlareAPIError(code, message)
CloudFlare.exceptions.CloudFlareAPIError: zonelockdown.api.unknown_error

**Comment by mahtin** [comment]
Try to run this with the added debug flag:
```
    cf = CloudFlare.CloudFlare(email=..., token=...,debug=True)
```
You will see the calls (in curl format) and the response info. The `zonelockdown.api.unknown_error` is coming back from the API call to Cloudflare, so you now have solved the errors you had before.

PS: It's three backquotes to add multi-line code in a markdown format. One backquote is useful for in-line code.

**Comment by anthonyd-cd** [comment]
Thanks @mahtin I added it but one thing I'm confused about is if I'm suppose to pass the whole config or not because when I tried this via the bash curl I was forced to pass both the urls and the ips associated with the lockdown rule, I got this error when I enabled debut

2021-02-26 22:38:31,077 - Python Cloudflare API v4 - DEBUG - Call: emulated curl command ...
       curl -X PUT "https://api.cloudflare.com/client/v4/zones/*********************************/firewall/lockdowns/*******************" \
            -H "User-Agent: python-cloudflare/2.8.15/python-requests/2.22.0/python/3.6.9" \
            -H "Content-Type: application/json" \
            -H "X-Auth-Email: admin.devops@canadadrives.ca" \
            -H "X-Auth-Key: REDACTED" \
            --data '['her.domain.ca/*', 'my.domain.ca/*', 'your.domain.ca/*', 'his.domain.ca/*', 'our.domain.ca/*']'
2021-02-26 22:38:31,166 - Python Cloudflare API v4 - DEBUG - Response: 400, application/json, b'{\n  "result": null,\n  "success": false,\n  "errors": [\n    {\n      "message": "zonelockdown.api.unknown_error"\n    }\n  ],\n  "messages": []\n}\n'

I'm guessing it wants it like this perhaps: 

'{"urls": '['her.domain.ca/*', 'my.domain.ca/*', 'your.domain.ca/*', 'his.domain.ca/*', 'our.domain.ca/*']', "configurations": [{"target": "ip_range", "value": "0.0.0.0/0"}]}, [{"target": "ip_range", "value": "::/0"}]}'

**Comment by mahtin** [comment]
OK - I got it - neither of us read the https://api.cloudflare.com/#zone-lockdown-update-lockdown-rule documentation correctly. I clearly states that `urls` and `configurations` are required parameters. It's my bad for not fully reading this before my first answer. So, here goes with the newest version of the code, I've cleaned up some variable names to make this clearer:
```
    data = cf.zones.firewall.lockdowns.get(zone_id)

    rule_number = 1
    fl_id = data[rule_number]['id']
    fl_urls = data[rule_number]['urls']
    fl_configurations = data[rule_number]['configurations']

    fl_urls.append('our.domain.ca')

    new_data = {'urls': fl_urls, 'configurations': fl_configurations}

    data_returned = cf.zones.firewall.lockdowns.put(zone_id, fl_id, data=new_data)
```
I've introduced `rule_number` (please make sure you mean `1` vs `0`). I've matched the `data` value in the `PUT` call to match the API documentation page (see url above). This is the key point! (and sorry once again for not reading that and realizing this error). I've also shown that the `PUT` call also returns the data values - should you care.

There is one other way to do this:
```
    data = cf.zones.firewall.lockdowns.get(zone_id)
    rule_number = 1
    data[rule_number]['urls'].append('our.domain.ca')
    data_returned = cf.zones.firewall.lockdowns.put(zone_id, fl_id, data=data)
``` 
In this example, you are just using the returned data structure and modifying it before sending it back. This also works (and is used in the README examples). However, both styles work - it's your choice.

One reason why you should consider using this second version is because this API also has optional values `paused` and `description`, which would be lost in the first version above but not the second. This could be fixed with:
```
    ...
    fl_paused = data[rule_number]['paused']
    fl_description = data[rule_number]['description']
    ...
    new_data = {'urls': fl_urls, 'configurations': fl_configurations, 'paused': fl_paused, 'description': fl_description}
```
As you can see, the second version (because it uses the original `data` response) always keeps these values intact.

I hope this helps. (I also note that the returned error values/strings are pretty useless in this case!)

**Comment by mahtin** [comment]
One more point (about the `rule_number` above). The call `GET zones/:zone_identifier/firewall/lockdowns` (see https://api.cloudflare.com/#zone-lockdown-list-lockdown-rules) and `GET zones/:zone_identifier/firewall/lockdowns/:id` (see https://api.cloudflare.com/#zone-lockdown-lockdown-rule-details) differ in that the first returns a `list` of all the rules vs just a single rule (and not a `list`).

Keep that in mind because your `PUT zones/:zone_identifier/firewall/lockdowns/:id` (see https://api.cloudflare.com/#zone-lockdown-update-lockdown-rule) only returns one value - the one you just updated. So in my example above `new_data` is not a list; but just one rule.

This is a very common setup within the Cloudflare API.

**Comment by anthonyd-cd** [comment]
Wow thanks a lot @mahtin ! thank you for having the patience to help me out!
I ended up going with the first method as the second one did not work and I had the same API error issue I initially did. 
I was able to get and put and ended up doing an if x in fl_urls: remove() in order to get it to work but tested it a few times and it works as expected! Thanks again and I hope to cross paths again

Stay safe out there! 👍
